### PR TITLE
adds async function for downloadData in Storage Plugin

### PR DESCRIPTION
**Pull request by brennanMKE**
*Issue #, if available:*

*Description of changes:*

*Check points: (check or cross out if not relevant)*

- [ ] Added new tests to cover change, if needed
- [ ] Build succeeds with all target using Swift Package Manager
- [ ] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [ ] PR title conforms to conventional commit style
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Amplify/Categories/Storage/Operation/StorageDownloadDataOperation.swift [hunk 1]
```diff
@@ -18,3 +18,7 @@ public extension HubPayload.EventName.Storage {
     /// eventName for HubPayloads emitted by this operation
     static let downloadData = "Storage.downloadData"
 }
+public typealias StorageDownloadDataTask = AmplifyInProcessReportingOperationTaskAdapter<StorageDownloadDataRequest,
+                                                                                                  Progress,
+                                                                                                  Data,
+                                                                                                  StorageError>
```
Comment (lawmicha) [inline Amplify/Categories/Storage/Operation/StorageDownloadDataOperation.swift:24]: line 22-24 alignment looks slightly off from first type "StorageDownloadDataRequest", ctrl+I ?

File: Amplify/Categories/Storage/StorageCategory+ClientBehavior.swift [hunk 2]
```diff
@@ -90,6 +90,12 @@ extension StorageCategory: StorageCategoryBehavior {
         try await plugin.getURL(key: key, options: options)
     }
 
+    @discardableResult
+    public func downloadData(key: String,
+                      options: StorageDownloadDataOperation.Request.Options?) async throws -> StorageDownloadDataTask {
+        try await plugin.downloadData(key: key, options: options)
```
Comment (lawmicha) [inline Amplify/Categories/Storage/StorageCategory+ClientBehavior.swift:96]: you may be confusing the protocol from this file, where the protocol doesn't have default parameters. StorageCategory` methods can and does have the `nil` default parameter for optional parameters, for example, the previous downloadData API on line 20 has the default parameter but this new async API does not. (https://github.com/aws-amplify/amplify-ios/issues/2223)

Comment (lawmicha) [reply Amplify/Categories/Storage/StorageCategory+ClientBehavior.swift:96]: nit: indentation

File: AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePlugin+AsyncClientBehavior.swift [hunk 3]
```diff
@@ -34,6 +34,21 @@ extension AWSS3StoragePlugin {
         return try await taskAdapter.value
     }
 
+    @discardableResult
+    public func downloadData(key: String,
+                      options: StorageDownloadDataOperation.Request.Options?) async throws -> StorageDownloadDataTask {
```
Comment (lawmicha) [inline AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePlugin+AsyncClientBehavior.swift:39]: indentation

File: AmplifyTests/CategoryTests/Hub/AmplifyOperationHubTests.swift [hunk 4]
```diff
@@ -228,6 +228,15 @@ class MockDispatchingStoragePlugin: StorageCategoryPlugin {
         return try await taskAdapter.value
     }
 
+    @discardableResult
+    public func downloadData(key: String,
+                      options: StorageDownloadDataOperation.Request.Options?) async throws -> StorageDownloadDataTask {
```
Comment (lawmicha) [inline AmplifyTests/CategoryTests/Hub/AmplifyOperationHubTests.swift:233]: indentation

**Review by lawmicha — APPROVED** [review]
LGTM w/ comments

File: Amplify/Categories/Storage/StorageCategory+ClientBehavior.swift (hunk 2 above)
Comment (brennanMKE) [reply Amplify/Categories/Storage/StorageCategory+ClientBehavior.swift:96]: yes, missed the indentation on this one. Once the async additions are in I'll be removing the listener functions. I can then make these other requested changes.

**Pull request merged**